doc/developer-notes.md

```diff
@@ -746,6 +746,72 @@ the upper cycle, etc.
 Threads and synchronization
 ----------------------------
 
+- Prefer `Mutex` type to `RecursiveMutex` one
+
+- Consistently use [Clang Thread Safety Analysis](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html) annotations to
+  get compile-time warnings about potential race conditions in code. Combine annotations in function declarations with
+  run-time asserts in function definitions:
+
+```C++
+// txmempool.h
+class CTxMemPool
+{
+public:
+    ...
+    mutable RecursiveMutex cs;
```

Comment:
nit: Maybe use an example that does not involve `RecursiveMutex` since we recommend against using it and if we manage to obliterate it from the code, then this example will require adjusting.

Reply:
Oh, I've failed to find an example in our code base that embraces a `Mutex` instance, an annotation in function member declaration, and an assertion in function member definition.